[merged segment warmer] support cleanup redundant pending merged segments

guojialiang92 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
This PR is based on [[18255](https://github.com/opensearch-project/OpenSearch/pull/18255)]'s follow-up work. It implements the core process of cleaning up redundant pending merged segments. Meanwhile, this cleanup process can be applied to both **local** and **remote storage** scenarios.

**Example**
Here, an example of generating redundant pending merged segments will be provided.

1. At time1, the primary shard merges `_1.si` and `_2.si` into segment `_3.si`. `_3.si` is pre-copied to the replica shard, which completed at time4.
2. At time2, the primary shard generated `_4.si`, and `_1.si`, `_2.si`, `_4.si` were copied to the replica shard via segment replication, which completed at time5.
3. At time3, the primary shard merges `_3.si` and `_4.si` into segment `_5.si`. `_5.si` is pre-copied to the replica shard, which completed at time6.
4. At time4, the replica completed the pre-copy of `_3.si`, but `_3.si` is not searchable.
5. At time5, the reference counts of `_1.si` and `_2.si` in the primary shard decreased to `0` and were deleted.
6. At time6, `_5.si` on the primary shard becomes searchable. The reference counts of `_3.si` and `_4.si` in the primary shard decreased to `0` and were deleted. `_5.si` were copied to the replica shard via segment replication, which completed at time7 (This process is fast because it has already performed pre-copy).
7. At time7, `_5.si` on the replica shard becomes searchable. The reference counts of `_1.si`, `_2.si` and `_4.si` in the replica shard decreased to `0` and were deleted.
8. After time7, `_3.si` on the replica is the redundant pending merged segment.

|     | primary | replica  |
| :------: | :--: | :----: |
| time1    | _1.si, _2.si, _3.si   |     |
| time2    | _1.si, _2.si, _3.si, _4.si   |     |
| time3    | _1.si, _2.si, _3.si, _4.si, _5.si   |     |
| time4    | _1.si, _2.si, _3.si, _4.si, _5.si   |  _3.si   |
| time5    | _3.si, _4.si, _5.si   |  _1.si, _2.si, _3.si, _4.si   |
| time6    | _5.si   |  _1.si, _2.si, _3.si, _4.si, _5.si   |
| time7    | _5.si   |  _3.si, _5.si   |

**Cleanup strategy**
When the following conditions are met, we will consider the pending merge segment to be redundant and can be deleted.

- The current primary term of the primary shard is greater than the primary term when the pending merge segment was generated, or the primary terms are identical but the segmentInfosVersion of the primary shard is greater than the segmentInfosVersion when the pending merge segment was generated.
- The pending merge segment no longer exists in the primary shard.

**Implement**

- After the merged segment pre-copy is completed, the replica shard needs to record `MergeSegmentCheckpoint` in `IndexShard#pendingMergeSegmentCheckpoints`.
- After the segment replication is completed, the replica shard needs to remove the relevant `MergeSegmentCheckpoint`  from `IndexShard#pendingMergeSegmentCheckpoints`.
- We introduced `IndexService.AsyncPublishReferencedSegmentsTask` to periodically publish the segments referenced by the primary shard to replicas.
- The replica shard cleans up the redundant pending merge segments according to the above **cleanup strategy**.



### Related Issues
Resolves #[[17528](https://github.com/opensearch-project/OpenSearch/issues/17528)], #[[18625](https://github.com/opensearch-project/OpenSearch/issues/18625)]
<!-- List any other related issues here -->

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Bukhtawar commented:
Thanks @guojialiang92 

> Example
Here, an example of generating redundant pending merged segments will be provided.
At time1, the primary shard merges _1.si and _2.si into segment _3.si. _3.si is pre-copied to the replica shard, which completed at time4.
At time2, the primary shard generated _4.si, and _1.si, _2.si, _4.si were copied to the replica shard via segment replication, which completed at time5.

Based on this example, I was wondering if _1.si and _2.si would actually be transferred at a higher priority to ensure minimal replication lag.

> At time4, the replica completed the pre-copy of _3.si, but _3.si is not searchable.

I am assuming soon after time4 the _3.si should be searchable.

guojialiang92 commented:
Thanks for taking a look @Bukhtawar. :)

Yeah, you are right. In order to reduce lag time, `_1.si` and `_2.si` will be transferred with higher priority based on segment replication. This is also one of the reasons why merged segment pre-copy decouples the merged segment from the incremental segment.
I'm just giving an example of redundant pending merged segment for everyone's easy understanding.

guojialiang92 commented:
> > At time7, _5.si on the replica shard becomes searchable. The reference counts of _1.si, _2.si and _4.si in the replica shard decreased to 0 and were deleted.
> 
> How does the reference cont of _1.si, _2.si and _4.si is getting decreased but not the _3.si file? Would you be able to shed more light on the stale file deletion happen as of today?

Hi, @ashking94. I'm glad to discuss with you. :)

Reference counting control in segment replication relies on `NRTReplicationEngine#replicaFileTracker`. In `NRTReplicationEngine#updateSegments` and `NRTReplicationEngine#flush`, the reference count will be incremented or decremented, and when the reference count of a file reaches 0, it will be deleted.

In the above example, `_1.si`, `_2.si`, and `_4.si` are sent from the primary shard to the replica shard through segment replication. At time7, `_5.si` completed segment replication and executed `NRTReplicationEngine#updateSegments` to update `NRTReplicationEngine#readerManager`. At this time, the reference count of `_5.si` increases by 1, and the reference count of `_1.si`, `_2.si`, and `_4.si` decreases by 1.

`_3.si` is sent from the primary shard to the replica through merged segment pre-copy and is not controlled by `NRTReplicationEngine#replicaFileTracker`.

A pending merged segment will only be managed by the above reference counting control when the checkpoint of segment replication includes that pending merged segment.

In the production environment, since the merge process is continuously ongoing, and segment replication always copies the latest checkpoint. This design based on eventual consistency results in some merged segments being deleted due to another merge operation on the primary shard before they have had time to be copied to replicas via segment replication. For these merged segments that have been deleted from the primary shard, in the current situation, replicas are unaware that they are no longer needed, thus resulting in redundancy.

To make it easier to understand, I have also added an IT test `MergedSegmentWarmerIT#testCleanupRedundantPendingMergeFile`, which creates multiple redundant segments.

Looking forward to your reply.

ashking94 commented:
@guojialiang92  Thanks for the clarification. I understand now why we need the unreferenced merged files to be cleaned up. I will review the PR shortly.

ashking94 commented:
>We introduced `IndexService.AsyncPublishReferencedSegmentsTask` to periodically publish the segments referenced by the primary shard to replicas.

Why is the publish action required?

ashking94 commented on server/src/test/java/org/opensearch/index/shard/SegmentReplicationIndexShardTests.java:
```diff
@@ -207,6 +208,52 @@ public void testMergedSegmentReplicationWithZeroReplica() throws Exception {
     }
 
+    @LockFeatureFlag(MERGED_SEGMENT_WARMER_EXPERIMENTAL_FLAG)
+    public void testCleanupRedundantPendingMergeSegment() throws Exception {
```
this is really a good test for simulating the scenario you mentioned in PR description.

guojialiang92 commented:
@ashking94 
> > We introduced `IndexService.AsyncPublishReferencedSegmentsTask` to periodically publish the segments referenced by the primary shard to replicas.
> 
> Why is the publish action required?

Since **segment replication** send segments from the primary shard to replicas, segments no longer referenced by the primary shard will not be sent to replicas in the future.

Here, the currently referenced segment is broadcast through the primary shard. If **pending merged segment** of the replica shard does not belong to these segments (primary referenced segments), it will be judged as redundant pending merged segment.

Of course, there can also be other cleaning mechanisms, and I can think of the following two.

1. When a replica receives a merged segment, it records a timestamp and deletes it based on an expiration time. This approach that solely relies on time may lead to misjudgment.
2. The replica sends a request to the primary shard to obtain the segments referenced by the primary shard. This approach has no fundamental difference from primary shard broadcast.

In summary, we used the publish action here.

ashking94 commented on server/src/test/java/org/opensearch/indices/replication/checkpoint/PublishReferencedSegmentsActionTests.java:
```diff
@@ -0,0 +1,230 @@
+
+        action.shardOperationOnPrimary(request, indexShard, ActionTestUtils.assertNoFailureListener(result -> {
+            // we should forward the request containing the current publish checkpoint to the replica
+            assertThat(result.replicaRequest(), sameInstance(request));
```
should we have some counter here that is validated outside to ensure that this block of code was definitely executed?

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1765,9 +1773,87 @@ public void finalizeReplication(SegmentInfos infos) throws IOException {
         if (engineOptional.isPresent()) {
             engineOptional.get().updateSegments(infos);
+            for (SegmentCommitInfo segmentCommitInfo : infos) {
+                String segmentCommitInfoName = segmentCommitInfo.info.name;
```
> I believe this will be the segment_N file?

Here we iterate through all segment names in `segmentInfos`, such as `_a`, `_b`, `_c`. If the segment name of the pending merged segment is among them, it will be removed from `pendingMergeSegmentCheckpoints`.

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1858,6 +1944,41 @@ ReplicationCheckpoint computeReplicationCheckpoint(SegmentInfos segmentInfos) th
+                if (file.startsWith(IndexFileNames.SEGMENTS) || file.startsWith("extra") || IndexWriter.WRITE_LOCK_NAME.equals(file)) {
+                    continue;
+                }
+                String segmentName = IndexFileNames.parseSegmentName(file);
```
Because we don't need to broadcast all files to replicas, only the segment names referenced by the primary shard need to be broadcast.

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1873,17 +1994,20 @@ public void publishMergedSegment(SegmentCommitInfo segmentCommitInfo) throws IOE
-            segmentMetadataMap,
-            segmentCommitInfo.info.name
-        );
+        try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = getSegmentInfosSnapshot()) {
+            SegmentInfos segmentInfos = new SegmentInfos(Version.LATEST.major);
+            segmentInfos.add(segmentCommitInfo);
+            Map<String, StoreFileMetadata> segmentMetadataMap = store.getSegmentMetadataMap(segmentInfos);
+            return new MergeSegmentCheckpoint(
+                shardId,
+                getOperationPrimaryTerm(),
+                segmentInfosGatedCloseable.get().getVersion(),
+                segmentMetadataMap.values().stream().mapToLong(StoreFileMetadata::length).sum(),
+                getEngine().config().getCodec().getName(),
+                segmentMetadataMap,
+                segmentCommitInfo.info.name
+            );
+        }
```
> This would be sending even the refreshed files that have been already replicated from earlier segrep occurrences. Can you explain what does this change do?

There's a bit of a misunderstanding here. We won't resend files that have already been sent. 
This change simply adds `segmentInfosGatedCloseable`'s version to the `MergeSegmentCheckpoint`.

ashking94 replied:
Got it.. I misread.

guojialiang92 commented on server/src/test/java/org/opensearch/index/shard/RemoteIndexShardTests.java:
```diff
@@ -645,6 +645,13 @@ public void testMergedSegmentReplicationWithZeroReplica() throws Exception {
 
+    @LockFeatureFlag(MERGED_SEGMENT_WARMER_EXPERIMENTAL_FLAG)
+    @Override
+    @AwaitsFix(bugUrl = "https://github.com/opensearch-project/OpenSearch/pull/18720")
```
> Why do we need to mute this?

The merged segment warmer of **remote store** has not been implemented yet, so this test is temporarily muted.

ashking94 commented:
@guojialiang92  Let me know once you have addressed all the comments. Will do the final review then.

guojialiang92 commented on server/src/main/java/org/opensearch/index/IndexService.java:
```diff
@@ -1182,6 +1198,14 @@ private void updateReplicationTask() {
     }
 
+    private void updatePublishReferencedSegmentsTask() {
+        try {
```
> only if the `publishReferencedSegmentsInterval` has changed

Already modified and added tests.

guojialiang92 commented on server/src/main/java/org/opensearch/index/IndexService.java:
```diff
@@ -1509,6 +1533,52 @@ private void maybeSyncSegments(boolean force) {
+                    shard.publishReferencedSegments();
+                }
+            } catch (IOException ex) {
+                // do nothing
```
done

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1765,9 +1773,87 @@ public void finalizeReplication(SegmentInfos infos) throws IOException {
+        pendingMergeSegmentCheckpoints.add(mergeSegmentCheckpoint);
+    }
+
+    public Set<MergeSegmentCheckpoint> getPendingMergeSegmentCheckpoints() {
```
done

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1858,6 +1944,41 @@ ReplicationCheckpoint computeReplicationCheckpoint(SegmentInfos segmentInfos) th
+     */
+    public ReferencedSegmentsCheckpoint computeReferencedSegmentsCheckpoint() throws IOException {
+        try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = getSegmentInfosSnapshot()) {
+            String[] allFiles = store.directory().listAll();
```
@ashking94 

> I am concerned that the directory may contain more files than current reader's view + merged segments. This can happen due to scroll or PIT search. Also, if there is any issue on primary, then it has tendency to now also happen on primary.

There are two parts here, the `listAll` operation and `parseSegmentName`.
1. `listAll`: During the segment replication process, the primary shard will execute `IndexShard#computeReplicationCheckpoint`, in which `Store#failIfCorrupted` will be called, followed by the `listAll` method. The frequency of "publish referenced segment" is much lower than that of "publish replication checkpoint", so I believe the I/O cost should be acceptable.
2. `parseSegmentName`: To address your concerns, we can only parse files with the `.si` suffix. Referring to the method in `Store#failIfCorrupted` that checks whether all files start with `corrupted_`, we can first check whether a file ends with `.si`. This can avoid a large number of invalid `parseSegmentName` operations.

Do you have any other suggestions?

ashking94 replied:
@guojialiang92 Can we remove `MergedSegmentCheckpoint` from `unSearchableMergedSegmentCheckpointList` only when it is searchable (which means that the `segmentInfosGatedCloseable` holds the file)? I was thinking if `publish_referenced_segments` action can act as that hook where we do the cleanup of segments that are present in `segmentInfosGatedCloseable` from `unSearchableMergedSegmentCheckpointList`. Also, around the memory leak point, we may have to take time information on when the checkpoint was added and clean it up after it is older than X mins/hours at the same time when the above cleanup is happening?

guojialiang92 replied:
@ashking94 

> Can we remove MergedSegmentCheckpoint from unSearchableMergedSegmentCheckpointList only when it is searchable (which means that the segmentInfosGatedCloseable holds the file)? 

Yes.

> I was thinking if publish_referenced_segments action can act as that hook where we do the cleanup of segments that are present in segmentInfosGatedCloseable from unSearchableMergedSegmentCheckpointList.

As mentioned above, we can introduce `searchableMergedSegmentCheckpointList`. 
The `computeReferencedSegmentsCheckpoint` can then use `searchableMergedSegmentCheckpointList` to add the pending merged segments files that are currently not part of `segmentInfosGatedCloseable`.

> Also, around the memory leak point, we may have to take time information on when the checkpoint was added and clean it up after it is older than X mins/hours at the same time when the above cleanup is happening?

Yes, but the default value may be quite large. Meanwhile, when users need to reduce the configuration, they will understand the risk that merged segments are no longer tracked, which may result in redundant pending merged segments on replica nodes not being cleaned up.

guojialiang92 commented on server/src/main/java/org/opensearch/indices/replication/MergedSegmentReplicationTarget.java:
```diff
@@ -84,6 +85,7 @@ protected void getFilesFromSource(
     @Override
     protected void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse) throws Exception {
         multiFileWriter.renameAllTempFiles();
+        indexShard.addPendingMergeSegmentCheckpoint((MergeSegmentCheckpoint) checkpoint);
```
> shall we add an assert here to check that the checkpoint is an instance of MergeSegmentCheckpoint?

done

guojialiang92 commented on server/src/main/java/org/opensearch/indices/replication/checkpoint/PublishReferencedSegmentsAction.java:
```diff
@@ -0,0 +1,100 @@
+            indexShard,
+            checkpoint,
+            new PublishReferencedSegmentsRequest(checkpoint),
+            "segrep_publish_referenced_segments",
```
done

guojialiang92 commented on server/src/test/java/org/opensearch/indices/replication/MergedSegmentReplicationTargetTests.java:
```diff
@@ -75,6 +75,7 @@ public void setUp() throws Exception {
             spyIndexShard.shardId(),
             spyIndexShard.getPendingPrimaryTerm(),
             1,
+            1,
```
> does the 1 have any relevance?

no, just for mock

guojialiang92 commented on server/src/test/java/org/opensearch/indices/replication/checkpoint/PublishReferencedSegmentsActionTests.java:
```diff
@@ -0,0 +1,230 @@
+            Sets.newHashSet("_1", "_2", "_3")
+        );
+
+        action.publish(indexShard, checkpoint);
```
In this unit test, we only ensure that doPublish can be called normally, and we leave the testing of the transport layer to IT test.

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1765,9 +1773,87 @@ public void finalizeReplication(SegmentInfos infos) throws IOException {
+        for (MergeSegmentCheckpoint mergeSegmentCheckpoint : pendingMergeSegmentCheckpoints) {
+            if (false == referencedSegmentsCheckpoint.getSegmentNames().contains(mergeSegmentCheckpoint.getSegmentName())
+                && referencedSegmentsCheckpoint.isAheadOf(mergeSegmentCheckpoint)) {
+                logger.trace(
```
> should we log both the objects and let the toString have the verbose string? basically `logger.trace("referencedSegmentsCheckpoint={}", referencedSegmentsCheckpoint)` once and `logger.trace("mergeSegmentCheckpoint={}", mergeSegmentCheckpoint)`?

done

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1765,9 +1773,87 @@ public void finalizeReplication(SegmentInfos infos) throws IOException {
+        List<MergeSegmentCheckpoint> pendingDeleteCheckpoints = new ArrayList<>();
+        for (MergeSegmentCheckpoint mergeSegmentCheckpoint : pendingMergeSegmentCheckpoints) {
+            if (false == referencedSegmentsCheckpoint.getSegmentNames().contains(mergeSegmentCheckpoint.getSegmentName())
+                && referencedSegmentsCheckpoint.isAheadOf(mergeSegmentCheckpoint)) {
```
@ashking94 
> This is what I understood - If the segment file which is present in pending merged segments is not present in referenced checkpoint, then we are cleaning it up if either published primary term is greater than the pending merge segment checkpoint or version is greater.
> 
> I have question - so, can it still be possible that segment infos version is lower while primary term is same, but it is still possible that the older segment infos is probably yet to be refreshed on the primary shard?

Your understanding is correct.
Let me explain this logic. The `version` and `term` information of both `MergedSegmentCheckpoint` and `ReferencedSegmentsCheckpoint` comes from the `segmentInfos` of the primary shard. The changes in `version` and `term` follow the following rules.

1. Under normal circumstances, the `term` does not change, and the `version` is monotonically increasing.
2. During the migration of the primary shard, the `term` does not change, segment replication occurs during the `hand-off` phase, and the `version` also increases monotonically.
3. When a replica is promoted to a primary shard, the `term` will increase.

In a design based on eventual consistency, redundant merged segment will eventually be deleted.

guojialiang92 commented:
@ashking94 
> @guojialiang92 Let me know once you have addressed all the comments. Will do the final review then.

I have addressed all the comments. Could you please help review it again? Thank you.

guojialiang92 commented:
Hi, @ashking94 
If you have time, please help review the code again. Thank you. :)

ashking94 reviewed (commented):
Thanks for the revision. I have some minor comments only now, let me know once you have addressed it.

ashking94 commented:
Do look at this - https://github.com/opensearch-project/OpenSearch/pull/18720#discussion_r2226002560 as well.

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -383,6 +386,8 @@ Runnable getGlobalCheckpointSyncer() {
     private final ClusterApplierService clusterApplierService;
     private final MergedSegmentPublisher mergedSegmentPublisher;
+    private final ReferencedSegmentsPublisher referencedSegmentsPublisher;
+    private final Set<MergedSegmentCheckpoint> pendingMergedSegmentCheckpoints = Sets.newConcurrentHashSet();
```
@ashking94 
> just to confirm - this is only being used on the replica?

Yes, `pendingMergedSegmentCheckpoints` is only used in replica shard.

guojialiang92 commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/MergedSegmentWarmerIT.java:
```diff
@@ -175,6 +182,87 @@ public void testMergeSegmentWarmerWithInactiveReplica() throws Exception {
+        primaryTransportService.addSendBehavior(
+            internalCluster().getInstance(TransportService.class, replicaNode),
+            (connection, requestId, action, request, options) -> {
+                if (action.equals(SegmentReplicationTargetService.Actions.FILE_CHUNK)) {
```
@ashking94 
> If we are failing the FILE_CHUNK transport action, wouldn't this be a problem for both refresh and merges? I didn't fully understand what we are doing by failing these calls.

Currently, we have two actions: `FILE_CHUNK` for segment replication and `MERGED_SEGMENT_FILE_CHUNK` for merged segment warmer. Therefore, if we failing the `FILE_CHUNK `transport action, it will only affect segment replication. We use this method to simply create redundant pending merged segments.

guojialiang92 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -1765,9 +1772,85 @@ public void finalizeReplication(SegmentInfos infos) throws IOException {
+     *
+     * @param referencedSegmentsCheckpoint primary referenced segments
+     */
+    public void cleanupRedundantPendingMergeSegment(ReferencedSegmentsCheckpoint referencedSegmentsCheckpoint) {
```
@ashking94 
> Do we need to handle for concurrency? Lets assume a case where there is a stale primary that is doing publish reference action and this method gets invoked almost in parallel? Lets see if this needs to be handled.

We ensure the order of `MergedSegmentCheckpoint` and `ReferencedSegmentsCheckpoint` through `ReplicationCheckpoint#isAheadOf`. 

Since `PublishReferencedSegmentsAction` is asynchronous, it is indeed possible for concurrent execution of `cleanupRedundantPendingMergeSegment` to occur. `pendingMergedSegmentCheckpoints` is of type `ConcurrentHashMap`, and concurrent processing will not cause exceptions, but due to weak consistency, it may result in traversing merged segments that have already been deleted.

ashking94 replied:
Sounds fine

guojialiang92 commented:
Hi, @ashking94 

> Thanks for the revision. I have some minor comments only now, let me know once you have addressed it.

Could you please help review it again? Including the discussion in [#18720 (comment)](https://github.com/opensearch-project/OpenSearch/pull/18720#discussion_r2226002560). 
Thank you.
